### DISCUSSION TITLE
The read operation timed out

### DISCUSSION THREAD

**Discussion by tv-ankur** [supabase/supabase#13977] [Questions] [answered]
Every time i try to call the function via the RPC , it gives me following error-

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/httpcore/backends/sync.py", line 28, in read
    return self._sock.recv(max_bytes)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 1263, in recv
    return self.read(buflen)
           ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/ssl.py", line 1136, in read
    return self._sslobj.read(len)
           ^^^^^^^^^^^^^^^^^^^^^^
TimeoutError: The read operation timed out
```

That's how I am calling the function-

  #search for the most similar transcripts
 ```
 matched_transcript = supabase.rpc('search_first_table',
                                    params={
                                      "query_embedding": query_embedding,
                                      "similarity_threshold": 0.70,
                                      "match_count": 20
                                    }).execute()
```

Here is how I have created 'search_first_table' function-

```
create or replace function search_first_table (
  query_embedding vector(1536),
  similarity_threshold float,
  match_count bigint
)
returns table (
  id bigint,
  subtitle_text text,
  start real,
  duration real,
  reference_video_id text,
  similarity float
)
language plpgsql
as $$
begin
  return query
  select
    first_table.id,
    first_table.subtitle,
    first_table.start,
    first_table.duration,
    first_table.video_id,
    1 - (first_table.subtitle_embedding <=> query_embedding) as similarity
  from first_table
  where 1 - (first_table.subtitle_embedding <=> query_embedding) > similarity_threshold
  order by first_table.subtitle_embedding <=> query_embedding
  limit match_count;
end;
$$;

```

This is how I have created the index-

```
create index on public.first_table 
using ivfflat (subtitle_embedding vector_cosine_ops)
with (lists = 100);
```

It was working perfectly for 200K records but as soon as I have added 250K records, it start giving me error. My plan to add more 400K records into the table. But this error keeping me away.

**Comment by gregnr** [comment]
Hey @tv-ankur! Sorry to hear about the timeout. Out of curiosity, when did you create the index?

`pgvector` recommends adding indexes only after the table has sufficient data, so that the internal IVFFlat cell clusters are based on your data's distribution. Anytime the distribution changes significantly, indexes should probably be recreated.

If you haven't already, could you try re-creating the indexes and see if that impacts the query?

**Reply by helloivanco** [reply]
@gregnr - I'm also running into a timeout issue but at SupabaseVectorStore.similaritySearchVectorWithScore (approximately 160k+ records). Following this example: https://js.langchain.com/docs/modules/data_connection/vectorstores/integrations/supabase which works fine when I tested with ~1000 records.

When I try to reindex with `reindex index concurrently embeddingstable_embedding_idx` in SQL editor it times out with a 504. What's the best way to perform the reindexing on large amounts of records? Is re-indexing likely to solve the vector similarity search timing out issue?

**Reply by gregnr** [reply]
Hey @helloivanco - since my last comment, pgvector has released an improved index type called HNSW:
https://supabase.com/blog/increase-performance-pgvector-hnsw

I would recommend switching over to this index type as it has a lot of improvements over IVFFlat, including faster query performance and no longer requiring you to re-index (even if you create it on 0 rows).

All new databases will automatically come with the latest pgvector (v0.5.0). If you have a database created prior to this, please create a [support request](https://supabase.com/dashboard/support/new) and we'll be happy to schedule an upgrade.

**Reply by helloivanco** [reply]
@gregnr - thank you for the amazingly fast response! 

I should have mentioned I have HNSW indexes setup:
<img width="761" alt="Screenshot 2023-09-30 at 1 46 13 AM" src="https://github.com/supabase/supabase/assets/70876305/57beb597-e7e9-400e-bdd7-ed28bb7fd198">

Vector 0.5.0 enabled:
<img width="381" alt="Screenshot 2023-09-30 at 1 46 43 AM" src="https://github.com/supabase/supabase/assets/70876305/1117f365-77bb-40d0-93ee-69a109a70a65">

Things were working fine with 1000 embedding records but after populating with 100k+, the SupabaseVectorStore.similaritySearchVectorWithScore started to time out.

**Reply by gregnr** [reply]
Got it - do you mind sharing your `match_documents()` function? I'm realizing there may be some improvements we need to make to that function so that it uses the index properly.

**Reply by helloivanco** [reply]
@gregnr - here is the match_documents function:

```
#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    1 - (documents.embedding <=> query_embedding) as similarity
  from documents
  where metadata @> filter
  order by documents.embedding <=> query_embedding
  limit match_count;
end;
```

The function is taken from this Langchain Supabase integration example: https://js.langchain.com/docs/modules/data_connection/vectorstores/integrations/supabase

**Reply by gregnr** [reply, marked as answer]
So turns out the function is fine - but I've realized your HNSW index is on the inner product operation (`vector_ip_ops`), but your `match_documents` function is using cosine distance (`<=>`). I would either switch your index to [`vector_cosine_ops`](https://supabase.com/docs/guides/ai/vector-indexes/hnsw-indexes#cosine-distance--vectorcosineops-), or re-write your function to use inner product (`<#>`). Note `<#>` is actually negative inner product, so you will have to negate your query appropriately.

**Reply by helloivanco** [reply]
@gregnr - thanks for pointing that out! I'll be switching the index to `vector_cosine_ops`. I've removed the previous index via the Supabase interface and trying to run `CREATE INDEX on documents USING hnsw (embedding vector_cosine_ops);
` but the statement times out while running it in the SQL editor. Is there a script or method to best re-index a very large Supabase table?

I've started a discussion thread on the issue I'm having: [Reindexing large table - how to avoid timeout errors?](https://github.com/orgs/supabase/discussions/17821)

**Reply by helloivanco** [reply]
@gregnr Started a fresh table with `vector_cosine_ops` indexing and keeping the cosine distance (<=>) match function . However, matching isn't returning relevant results. It was oddly matching better with `vector_ip_ops' and the cosine distance match function. Here are the current index, match function and store embeddings:

Index
<img width="735" alt="Screenshot 2023-10-03 at 1 38 05 PM" src="https://github.com/supabase/supabase/assets/70876305/7fc8d67e-841b-4519-b0df-e90d9298701e">

Match function
<img width="637" alt="Screenshot 2023-10-03 at 1 38 33 PM" src="https://github.com/supabase/supabase/assets/70876305/45eaefc1-4987-48a5-9372-2ea65bc80508">

Store embeddings
<img width="635" alt="Screenshot 2023-10-03 at 1 46 44 PM" src="https://github.com/supabase/supabase/assets/70876305/e5a963ec-89ef-49a6-b1a1-a73671c28c21">

Also seeing another project use `vector_ip_ops' and the cosine distance match function.
https://github.com/kaarthik108/snowChat/blob/main/supabase/scripts.sql

**Reply by gregnr** [reply]
Hey @helloivanco - that's strange. Can you confirm whether or not the results are still relevant without any index created?

**Reply by helloivanco** [reply]
Hi @gregnr - it seems to work fine after updating the match function.

With index `CREATE INDEX ON documents USING hnsw (embedding vector_cosine_ops);` this updated match function seems to work fine so far (about ~60k embeddings). 

```
-- Create a function to search for documents
create function match_documents (
  query_embedding vector(1536),
  match_count int DEFAULT 3,
  match_threshold float DEFAULT 0.78,
  filter jsonb DEFAULT '{}'
) returns table (
  id bigint,
  content text,
  metadata jsonb,
  similarity float
)
language plpgsql
as $$
#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    1 - (documents.embedding <=> query_embedding) as similarity
  from documents
  where metadata @> filter and 1 - (documents.embedding <=> query_embedding) > match_threshold
  order by similarity desc
  limit match_count;
end;
$$;
```

**Reply by gregnr** [reply]
@helloivanco glad to hear this is working - though I'm concerned that your new function might miss the index:

My query plan when using `1 - distance` appears to miss the index and fall back on a sequential scan:

![image](https://github.com/supabase/supabase/assets/4133076/33ffa398-3a43-46df-a635-4a2a6d4d56ca)

As opposed to ordering by just `distance`, we get an index scan:

![image](https://github.com/supabase/supabase/assets/4133076/c7439377-9974-442b-bca8-4d43ebc41ca8)

These differences would apply when the calculation is used in either the `where` clause or `order by` clause (but not `select`, since those are only calculated after the results have been filtered/ordered).

TBH, I'm surprised that your old function produced different results at all since `1 - distance` descending should produce the same rows as `distance` ascending.

Some things to consider
- Are your embeddings normalized? If not, `1 - distance` wouldn't produce the correct similarity
- Were you using the same limits on both your old and new functions?
- How is the performance of your new match function with and without the index? If it's identical, it's likely not using the index.
- You could try running `explain` yourself to see if your query plan is using the index

**Reply by helloivanco** [reply]
@gregnr - that's really helpful. I ran `explain` and got the same results as you and have updated the function to take advantage of the indexing. At about 100k+ embedding records, sequential scan is not possible and starts to to timeout (my original issue). 

While using the old function (sequential scan) the results did appear to better match but it could also be due less embeddings in the table to compare to at the time. Need to investigate this further. 

So here is what I'm using as the match function now: 
```
-- Create a function to search for documents -- version 2
create function match_documents (
  query_embedding vector(1536),
  match_count int DEFAULT 6,
  match_threshold float DEFAULT 0.16,
  filter jsonb DEFAULT '{}'
) returns table (
  id bigint,
  content text,
  metadata jsonb,
  similarity float
)
language plpgsql
as $$
#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    1-(documents.embedding <=> query_embedding) as similarity
  from documents
  where (documents.embedding <=> query_embedding) < match_threshold
  order by documents.embedding <=> query_embedding
  limit match_count;
end;
$$;
```
_For those running the `explain` test. Here is the generated_normalized_array function I used:_
```
-- generate normalized array
create or replace function generate_normalized_array(dimension int)
	returns double precision[]
    language sql
as $$
    with unnormed(elem) as(
        select random() from generate_series(1, $1) v(x)
    ),
    norm(factor) as (
        select
            sqrt(sum(pow(elem, 2)))
        from
            unnormed
    )
    select
        array_agg(u.elem / r.factor)
    from
        unnormed u, norm r
$$;
```

**Reply by HexNinja555** [reply]
Hello, this looks like a classic scaling + indexing issue with pgvector.

Since it worked at ~200K rows and started timing out after ~250K, the problem is likely not your function, but how the index is being used (or not used).

A few things to check:

1. Index vs operator mismatch

   Make sure your index matches your similarity operator.
   Right now you're using cosine distance (`<=>`) in your function, so your index must be:

```sql
using ivfflat (subtitle_embedding vector_cosine_ops)
```

If there's any mismatch (like `vector_ip_ops`), Postgres won’t use the index properly and will fall back to a slow scan.

2. Recreate the index after large data changes
   IVFFlat indexes depend on data distribution. If you added ~250K rows after creating the index, it can become inefficient.
   Try:

```sql
drop index if exists first_table_subtitle_embedding_idx;

create index on public.first_table 
using ivfflat (subtitle_embedding vector_cosine_ops)
with (lists = 100);
```

3. Check if the index is actually used
   Run:

```sql
explain analyze
select ...
```

If you see a sequential scan → that’s your bottleneck.

4. Increase lists for larger datasets
   For ~500K+ rows, try:

```sql
with (lists = 200 or 300)
```

More lists = better recall, faster queries (but slightly more memory).

5. Try HNSW if possible
   If you're on newer pgvector:

```sql
create index on first_table using hnsw (subtitle_embedding vector_cosine_ops);
```

HNSW tends to scale better for large datasets and avoids some IVFFlat tuning issues.

6. Timeout might be client-side
   Since the error is:

```
TimeoutError: The read operation timed out
```

you may also want to increase your client timeout or test the query directly in SQL to confirm.

---

TL;DR:

* Ensure operator + index match (`<=>` + `vector_cosine_ops`)
* Rebuild index after large inserts
* Tune `lists`
* Verify index usage with EXPLAIN
* Consider switching to HNSW

Let me know what EXPLAIN shows — that’ll tell us exactly where the slowdown is 👍

**Comment by tv-ankur** [comment]
@gregnr  thanks for answering my query.. 

I have created the index when there was no data. 

When I try to again create the index- it gives following error-

<img width="669" alt="Screenshot 2023-04-26 at 11 33 39 AM" src="https://user-images.githubusercontent.com/118169613/234483642-e2c2487e-8390-482a-a698-d7f4da2e41aa.png">

**Reply by gregnr** [reply]
Hmm, that's a odd one. Do you continue to get that error after logging out and back in?

**Reply by tv-ankur** [reply]
> Hmm, that's a odd one. Do you continue to get that error after logging out and back in?

Yes.. @gregnr

**Reply by tv-ankur** [reply]
@gregnr any updates?

**Reply by gregnr** [reply]
Hey @tv-ankur, sorry for the delay. Haven't figured this one out yet - doesn't seem to be any reports from other users. Can you please confirm if you're able to run another query successfully? You can try:
```sql
select 1;
```

**Reply by tv-ankur** [reply]
@gregnr yes, I am able to run other queries. Do you want to take look into my account?

**Reply by jon-qualifai** [reply]
I am also seeing this error when trying to read from a Vector table. Funny timing when I saw this thread!

**Reply by tv-ankur** [reply]
> I am also seeing this error when trying to read from a Vector table. Funny timing when I saw this thread!

@jon-qualifai hahah.. are you also seeing when you add more data? or its happening even with less than 100K records?

**Reply by jon-qualifai** [reply]
I was seeing this error when reading from the Vector table via the python-sdk
![image](https://user-images.githubusercontent.com/121834267/234854334-83c89717-e510-4b7e-ac29-18d277c84647.png)

**Reply by jon-qualifai** [reply]
I've tried changing the default timeouts but doesn't seem to make a difference.
https://supabase.com/docs/guides/database/timeouts#changing-the-default-timeout

**Reply by tv-ankur** [reply]
> I was seeing this error when reading from the Vector table via the python-sdk ![image](https://user-images.githubusercontent.com/121834267/234854334-83c89717-e510-4b7e-ac29-18d277c84647.png)

Ahh you are fetching via the table function, but I was fetching data via the RPC. One trick you can do is- lower the limit and specify which field you need. 

It look like python sdk have lower timout time.

**Reply by tv-ankur** [reply]
> I've tried changing the default timeouts but doesn't seem to make a difference. https://supabase.com/docs/guides/database/timeouts#changing-the-default-timeout

Just my observation, I may be wrong-- this only set the timeout in the browser session. But calling the table via the SDK is in not in same session. That's why it don't work when you call table via SDK. 

@jon-qualifai

**Reply by jon-qualifai** [reply]
Got it, thanks, useful to know RPC is an option.

**Reply by tv-ankur** [reply]
> Got it, thanks, useful to know RPC is an option.

If you can use, use JS SDK rather than Python. It have more features and stable than python one.

**Reply by gregnr** [reply]
@jon-qualifai Thanks for reporting. How many records are you working with on your `embeddings` table and do you have any indexes?

@tv-ankur I'm starting to think the CORS error is just a symptom of another issue. Can you confirm the status code of the above response that errored? Can you also provide the query that you're running that produces this error? Thanks!

By the way, I assume you are running the query in the studio editor? One work around in the mean time you can try is connecting to your Postgres DB directly, ie. via `psql` CLI, then try running the same query there. If you try this out, let me know!

**Reply by seveelly** [reply]
the same issue is with the javascript client ithrows {
  "code": "XX000",
  "details": null,
  "hint": null,
  "message": "Operation timed out after 5001 milliseconds with 0 bytes received"
}

**Reply by GaryAustin1** [reply]
@seveelly Are you sure yours is related?  5seconds happens to be the timeout for the http extension.  Any chance that is involved in your code?  You really should show some code for help.

**Reply by seveelly** [reply]
how can you fix http extension timeout

**Reply by seveelly** [reply]
i have tried SELECT http_set_curlopt('CURLOPT_TIMEOUT', '20');

**Reply by GaryAustin1** [reply]
This is really not the topic for this, but so far myself and 2 other users have not figured out how to get the time extended...
https://github.com/orgs/supabase/discussions/13760

**Reply by jon-qualifai** [reply]
We're still having this issue, anyway to increase timeouts when calling from Python? they are very short and throwing errors if a request takes more than a few seconds.

**Reply by GaryAustin1** [reply]
Assuming you are not talking about the http timeout (which was off topic)
https://supabase.com/docs/guides/database/timeouts

**Reply by jon-qualifai** [reply]
We've tried adjusting those settings 

`alter role <role> set statement_timeout ='60s';`
and 
`set statement_timeout to 60000; commit; -- 1 minute in milliseconds`

but still get timeouts when calling Supabase from the Python SDK after a few seconds.

**Reply by GaryAustin1** [reply]
https://github.com/supabase-community/postgrest-py/issues/72
Maybe related.  Don't use python, but it seems to have its own timeout.

**Reply by ishaqmahsud** [reply]
httpx has timeout which think 4 secs, due to the error arise "timeout read operation", in js library it work fine but in python it scary. i think team needs it urgent basis to resolve this issue.

**Reply by GaryAustin1** [reply]
@ishaqmahsud this topic started on the http postgres extension. Although Python is mentioned later it is not relevant to the original question.  If you feel the setup is not correct for Python then you should generate an issue in https://github.com/supabase-community/supabase-py or in the main Supabase issues section. Otherwise start a new discussion if you are seeking user help.

**Comment by JimmyLv** [comment]
My `documents` table also always throws a '57014' error, this table is what I use to store OpenAI embedding vectors, and it's currently quite large. 

Is there any documentation on optimization for this?

```ts
import supabaseClient from '@/lib/supabase/client'

export async function getVideoDocumentsByUrl(url: string) {
  const { data: existingContent, error } = await supabaseClient
    .from('documents')
    .select('content, metadata')
    .eq('metadata->>source', url)

  if (error) {
    console.error('Failed to getVideoDocumentsByUrl', error)
    throw error
  }
  return existingContent
}
```

**Reply by HexNinja555** [reply]
Hello, that `57014` error is usually a **query timeout / statement cancellation**, not a bug in your code.

Since your `documents` table is large and you're querying on:

```sql
metadata->>'source'
```

the issue is very likely **missing indexes**, which causes a full table scan.

### 🔍 What’s happening

Postgres can’t use a normal index for `metadata->>'source'` unless you explicitly create one.
So with large embedding tables → it scans everything → times out.

---

### ✅ Fix (this is the important part)

Create an index on that JSON field:

```sql
create index idx_documents_source
on documents ((metadata->>'source'));
```

👉 This alone usually fixes the timeout.

---

### 🚀 If you're using filters often (better approach)

Use a GIN index for JSONB:

```sql
create index idx_documents_metadata_gin
on documents using gin (metadata);
```

Then rewrite query (if possible):

```sql
where metadata @> '{"source": "your_url"}'
```

👉 Much faster for large datasets.

---

### 🧠 Bonus (for vector + large tables)

Since you're also storing embeddings:

1. Keep metadata queries fast (index like above)
2. Avoid mixing heavy vector search + JSON filtering in same query if possible
3. If needed, split:

   * Step 1: filter IDs by metadata
   * Step 2: run vector search

---

### ⚡ Quick debug tip

Run:

```sql
explain analyze
select content, metadata
from documents
where metadata->>'source' = 'your_url';
```

If you see **Seq Scan** → index is missing.

---

### TL;DR

* `57014` = query timeout
* Root cause = no index on `metadata->>'source'`
* Fix = create expression index or GIN index

---

That should remove the timeout completely 👍
